pallets/tokens/src/lib.rs

```diff
@@ -875,3 +911,258 @@ where
+		let actual = Module::<T>::slash_reserved(currency_id, who, value);
+		(MultiTokenNegativeImbalance::zero(currency_id), actual)
+	}
+
+	fn reserved_balance(currency_id: T::CurrencyId, who: &T::AccountId) -> Self::Balance {
+		Module::<T>::reserved_balance(currency_id, who)
+	}
+
+	fn reserve(currency_id: T::CurrencyId, who: &T::AccountId, value: Self::Balance) -> DispatchResult {
+		Module::<T>::reserve(currency_id, who, value)
+	}
+
+	fn unreserve(currency_id: T::CurrencyId, who: &T::AccountId, value: Self::Balance) -> Self::Balance {
+		Module::<T>::unreserve(currency_id, who, value)
+	}
+
+	fn repatriate_reserved(
+        currency_id: T::CurrencyId,
+		slashed: &T::AccountId,
+		beneficiary: &T::AccountId,
+		value: Self::Balance,
+		status: Status,
+	) -> result::Result<Self::Balance, DispatchError> {
+		Module::<T>::repatriate_reserved(currency_id, slashed, beneficiary, value, status)
+	}
+}
+
+impl<T> MultiTokenLockableCurrency<T::AccountId> for MultiTokenCurrencyAdapter<T>
+where
+	T: Trait,
+{
+	type Moment = T::BlockNumber;
+	type MaxLocks = ();
+
+	fn set_lock(currency_id: T::CurrencyId, id: LockIdentifier, who: &T::AccountId, amount: Self::Balance, _reasons: WithdrawReasons) {
+		Module::<T>::set_lock(id, currency_id, who, amount)
+	}
+
+	fn extend_lock(currency_id: T::CurrencyId, id: LockIdentifier, who: &T::AccountId, amount: Self::Balance, _reasons: WithdrawReasons) {
+		Module::<T>::extend_lock(id, currency_id, who, amount)
+	}
+
+	fn remove_lock(currency_id: T::CurrencyId, id: LockIdentifier, who: &T::AccountId) {
+		Module::<T>::remove_lock(id, currency_id, who)
+	}
+}
+
+
+impl<T> MultiTokenCurrencyExtended<T::AccountId> for MultiTokenCurrencyAdapter<T> where
+    T: Trait
+{
+	fn create(address: &T::AccountId, amount: T::Balance) -> T::CurrencyId{
+		let token_id = <NextCurrencyId<T>>::get();
+		<NextCurrencyId<T>>::mutate(|id| *id += One::one());
+		// we are creating new token so amount can not be overflowed as its always true
+		// 0 + amount < T::Balance::max_value()
+		let _ = <Self as MultiTokenCurrency<T::AccountId>>::deposit_creating(token_id, address, amount);
+		token_id
+	}
+
+	fn mint(currency_id: T::CurrencyId, address: &T::AccountId, amount: T::Balance) -> DispatchResult {
+		if ! Self::exists(currency_id){
+			return Err(DispatchError::from(Error::<T>::TokenIdNotExists));
+		}
+		let current_balance =  <Self as MultiTokenCurrency<T::AccountId>>::total_balance(currency_id, address);
+		// check for overflow while minting
+		current_balance.checked_add(&amount).ok_or(Error::<T>::BalanceOverflow)?;
+		let tokens = <Self as MultiTokenCurrency<T::AccountId>>::issue(currency_id, amount);
+		// we are always sure that there will be no overflow as it was checked
+		let _ = <Self as MultiTokenCurrency<T::AccountId>>::resolve_into_existing(currency_id, address, tokens);
+		Ok(())
+	}
+
+	fn exists(currency_id: Self::CurrencyId) -> bool{
+		<TotalIssuance<T>>::contains_key(currency_id)
+	}
+
```

Comment:
burning should be only performing free balance - maybe burn function should be introduced ??